crates/ide_completion/src/item.rs

```diff
@@ -354,12 +354,13 @@ pub(crate) struct Builder {
 }
 
 impl Builder {
-    pub(crate) fn build(self) -> CompletionItem {
+    pub(crate) fn build(&self) -> CompletionItem {
```

Comment:
this method should remain as is, it shoud consume the builder and turn it into the CompletionItem.

Reply:
In order to keep this signature, I need to `clone` the builder before calling `build` or I'm getting this error:
![image](https://user-images.githubusercontent.com/67855609/110798235-9e696a00-8282-11eb-9ba6-91b8e5098b31.png)
Do you think its better that way?

Reply:
`Buider` can't implement the `Copy` trait because of all the `Option`s inside it

Reply:
don't write `let builder = &mut CompletionItem::new`, use `let mut builder = CompletionItem::new()`

Reply:
Thank you for helping :smiley: 
This is not the case here, please notice that `builder` is initiated without `&mut`:
![image](https://user-images.githubusercontent.com/67855609/110800138-93afd480-8284-11eb-8380-67e551a0e236.png)

This happens because `kind`, `detail` etc. now returns `&mut`, so `item` is `&mut`. In order to call a function with `self` one must have an owned instance of this variable, and can't use a reference.
As far as I understand, the solutions are either changing `build` to accept a `&self` or `clone`ing the builder before building. the latter sounds worst in my opinion.

Reply:
You don't need `let mut item = builder().some_method()`. Use the same builder everywhere, don't reassign it.

Reply:
Doing this raises `creates a temporary which is freed while still in use`... I must give it a variable so that it will not get freed.
:disappointed:

Reply:
the following works for me

```
    fn add_field(&mut self, field: hir::Field, ty: &Type) -> CompletionItem {
        let is_deprecated = self.ctx.is_deprecated(field);
        let name = field.name(self.ctx.db());
        let mut builder = CompletionItem::new(
            CompletionKind::Reference,
            self.ctx.source_range(),
            name.to_string(),
        );
        builder
            .kind(SymbolKind::Field)
            .detail(ty.display(self.ctx.db()).to_string())
            .set_documentation(field.docs(self.ctx.db()))
            .set_deprecated(is_deprecated);

        if let Some(relevance) = compute_relevance(&self.ctx, &ty, &name.to_string()) {
            builder.set_relevance(relevance);
        }

        builder.build()
    }
```

Reply:
Oh! you also changed `builder = builder.set_relevance(relevance);` to `builder.set_relevance(relevance);` (without reassigning), wasn't shore it was ok.
Thank you, I'll change that.

Reply:
> wasn't shore it was ok.

Don't be afraid to make decisions as contributor :) The goal is to produce a code that makes sense  and solves the problem, the *specific* sequence of characters is never the final target.

Reply:
Well said :) I'm more happy with the results now, please re-review and let me know if there's anything else.